pkg/ottl/parser.go

```diff
@@ -44,13 +55,21 @@ type Statement[K any] struct {
 func (s *Statement[K]) Execute(ctx context.Context, tCtx K) (any, bool, error) {
 	condition, err := s.condition.Eval(ctx, tCtx)
 	if err != nil {
-		return nil, false, err
+		if s.errorMode == PropagateError {
+			return nil, false, err
+		}
+		s.telemetrySettings.Logger.Error("error executing condition", zap.Error(err))
+		return nil, false, nil
```

Comment:
This to me looks like unnecessary here. The caller can decide what to do with the error. If a statement fail at any point there is not like we can continue executing that statement, so we need to stop and I would say return an error.

The caller of `Execute` can determine what they want to do.

I fail to understand where this extra code is useful for a caller?

Reply:
I would see this useful in a "type Statements struct {statements []Statement}" that executes multiple statements to ignore the error from a previous statement and continue with next one.

Reply:
@bogdandrutu I believe main advantage of adding error handling like this to the OTTL is that is allows us to set some expectations when using OTTL.  

By including error handling logic to OTTL it becomes much easier to enforce the concept of "open erroring" within the package itself, which then allows us to make more exact statements about how functions should handle errors.  By including this type of error handling we can make a statement like "functions should always return any failed type assertions as errors", while ensuring that OTTL will gracefully handle the error. 

It also has the benefit of providing a default error handling strategy for components that use OTTL.  If they want to always keep processing, they don't have to do anything extra.

We don't provide a `type Statements struct {statements []Statement}` struct, but I can see how this error handling fits nicely with that struct. I'll play with that as part of this PR and update the implementing components to use it in a future PR.  

A struct like that would help with the future problem I foresee where an implementing component, like the transformprocessor, wants to configure not only the ability to skip to the next statement on error, but also skip to the next TransformContext and halting (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16519#issuecomment-1405425167)

Reply:
@bogdandrutu add new `Statements` struct and moved the error handling there.  In a followup PR I''l update `ParseStatements` to return a `Statements` struct.  This will work for all of the components in contrib that use OTTL except for routingprocessor.  For the routingprocessor, which only needs to execute a single statement and wants the results, condition, and error, I'll add a new `ParseStatement` function that just returns a `Statement`.  What do you think?

Reply:
Since `Parser` will end up with the ability to parse a list of statements and a single statement, I think `ErrorMode` will be more appropriate on `ParseStatements` instead of the whole `Parser` struct.

Reply:
If anyone is curious, I played around with the result and I like the implications: https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a6c2f19a6988a167dde7eb054867e1d0223be203